app/src/main/java/com/github/libretube/api/poToken/PoTokenWebView.kt

```diff
@@ -0,0 +1,279 @@
+            }
+        }
+    }
+
+    /**
+     * Called by the JavaScript snippet from [generatePoToken] when an error occurs in calling the
+     * JavaScript `obtainPoToken()` function.
+     */
+    @JavascriptInterface
+    fun onObtainPoTokenError(identifier: String, error: String) {
+        if (BuildConfig.DEBUG) {
+            Log.e(TAG, "obtainPoToken error from JavaScript: $error")
+        }
+        poTokenContinuations.remove(identifier)?.resumeWithException(PoTokenException(error))
+    }
+
+    /**
+     * Called by the JavaScript snippet from [generatePoToken] with the original identifier and the
+     * result of the JavaScript `obtainPoToken()` function.
+     */
+    @JavascriptInterface
+    fun onObtainPoTokenResult(identifier: String, poTokenU8: String) {
+        if (BuildConfig.DEBUG) {
+            Log.d(TAG, "Generated poToken (before decoding): identifier=$identifier poTokenU8=$poTokenU8")
+        }
+        val poToken = try {
+            u8ToBase64(poTokenU8)
+        } catch (t: Throwable) {
+            poTokenContinuations.remove(identifier)?.resumeWithException(t)
+            return
+        }
+
+        if (BuildConfig.DEBUG) {
+            Log.d(TAG, "Generated poToken: identifier=$identifier poToken=$poToken")
+        }
+        poTokenContinuations.remove(identifier)?.resume(poToken)
+    }
+
+    fun isExpired(): Boolean {
+        return Instant.now().isAfter(expirationInstant)
+    }
+    //endregion
+
+    //region Utils
+    /**
+     * Makes a POST request to [url] with the given [data] by setting the correct headers.
+     * This is supposed to be used only during initialization. Returns the  response body
+     * as a String if the response is successful.
+     */
+    private suspend fun makeBotguardServiceRequest(url: String, data: List<String>): String = withContext(Dispatchers.IO) {
+        val response = RetrofitInstance.externalApi.botguardRequest(url, data)
+        response.toString()
+    }
+
+    /**
+     * Handles any error happening during initialization, releasing resources and sending the error
+     * to [generatorContinuation].
+     */
+    private fun onInitializationError(error: Throwable) {
+        CoroutineScope(Dispatchers.Main).launch {
+            close()
+            generatorContinuation.resumeWithException(error)
+        }
+    }
+
+    /**
+     * Releases all [webView] resources.
+     */
+    @MainThread
+    fun close() = with(webView) {
+        clearHistory()
+        // clears RAM cache and disk cache (globally for all WebViews)
+        clearCache(true)
+
+        // ensures that the WebView isn't doing anything when destroying it
+        loadUrl("about:blank")
+
+        onPause()
+        removeAllViews()
+        destroy()
+    }
+    //endregion
+
+    companion object {
+        private val TAG = PoTokenWebView::class.simpleName
+        private const val REQUEST_KEY = "O43z0dpjhgX20SCx4KAo"
+        private const val JS_INTERFACE = "PoTokenWebView"
```

Comment:
I assume that this is just the class name, so we could in theory also do `PoTokenWebView::class.simpleName!!` here instead of hardcoding?